### PR TITLE
SB-930: OrientDB cascade update

### PR DESCRIPTION
**Pull request by sbespalov**

### CODE REVIEW

**Comment by fuss86** [comment]
I'll take a look on it over the weekend if it's not too late :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java [hunk 1]
```diff
@@ -57,6 +57,8 @@
     public <S extends ArtifactEntry> S save(S entity,
                                             boolean updateLastVersion)
     {
+        entity.setArtifactCoordinates(entity.getArtifactCoordinates());
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java]: self replacement

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java [hunk 2]
```diff
@@ -39,29 +47,99 @@ public void postConstruct()
...
+                    ReflectionUtils.setField(field, entity, newFieldValue);
+                }
+            }
+
         });
+
+        return getDelegate().save(entity);
     }
 
+    private Set<CascadeType> exposeCascadeType(Annotation a)
```
Comment (fuss86) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java]: Consider new class for this utility method

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java [hunk 3]
```diff
@@ -39,29 +47,99 @@ public void postConstruct()
...
+            Object fieldValue = ReflectionUtils.getField(field, entity);
+
+            if (fieldValue == null)
             {
                 return;
             }
 
-            CommonCrudService<GenericEntity> entityService = (CommonCrudService<GenericEntity>) entityServiceRegistry.getEntityService(subEntity.getClass());
-            entityService.identifyEntity(subEntity);
+            if (Collection.class.isAssignableFrom(fieldType))
```
Comment (fuss86) [inline strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java]: Will we support `Map`-s ? For example, `org.carlspring.strongbox.configuration.Configuration#storages`

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserAccessModel.java [hunk 4]
```diff
@@ -19,6 +21,7 @@
 {
 
     @XmlElement
+    @OneToOne(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserAccessModel.java]: Don't think it is necessary since `UserAccessModel` is a field of `org.carlspring.strongbox.security.User` and this class is not a `GenericEntity`. So `UserAccessModel` shouldn't be a `GenericEntity` too, I guess.

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserPathPermissions.java [hunk 5]
```diff
@@ -21,6 +24,7 @@
 {
 
     @XmlElement(name = "path")
+    @OneToMany(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserPathPermissions.java]: as above, probably unnecessary+ `UserPathPermissions` should not be a `GenericEntity`, unless I've missed something

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserRepositories.java [hunk 6]
```diff
@@ -21,6 +24,7 @@
 {
 
     @XmlElement(name = "repository")
+    @OneToMany(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserRepositories.java]: as above

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserRepository.java [hunk 7]
```diff
@@ -27,6 +29,7 @@
     private String repositoryId;
 
     @XmlElement(name = "path-permissions")
+    @OneToOne(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserRepository.java]: as above

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserStorage.java [hunk 8]
```diff
@@ -20,6 +22,7 @@
 {
 
     @XmlElement(name = "repositories")
+    @OneToOne(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserStorage.java]: as above

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserStorages.java [hunk 9]
```diff
@@ -21,6 +24,7 @@
 {
 
     @XmlElement(name = "storage")
+    @OneToMany(cascade = CascadeType.ALL)
```
Comment (fuss86) [inline strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserStorages.java]: as above

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/AccessModel.java [hunk 10]
```diff
@@ -21,7 +21,6 @@
  * @author Alex Oreshkevich
  */
 public class AccessModel
-        extends GenericEntity
```
Comment (fuss86) [inline strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/AccessModel.java]: I'm curious why is this changed ? `AccessModel` is a persistent field of entity `org.carlspring.strongbox.users.domain.User`

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java [hunk 11]
```diff
@@ -26,6 +30,7 @@
 
     private String securityTokenKey;
 
+    @Embedded
```
Comment (fuss86) [inline strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: Shouldn't `AccessModel` be `Embeddable` in this case ? When do we support this annotation in our codebase ?

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/security/AuthorizationConfig.java [hunk 12]
```diff
@@ -3,7 +3,9 @@
 import org.carlspring.strongbox.data.domain.GenericEntity;
 import org.carlspring.strongbox.security.Roles;
 
+import javax.persistence.CascadeType;
```
Comment (fuss86) [inline strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/security/AuthorizationConfig.java]: unnecessary import

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/security/AuthorizationConfig.java [hunk 13]
```diff
@@ -3,7 +3,9 @@
...
 import javax.persistence.Embedded;
+import javax.persistence.OneToOne;
```
Comment (fuss86) [inline strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/security/AuthorizationConfig.java]: same as above

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java [hunk 14]
```diff
@@ -52,7 +54,9 @@
     private LayoutProviderRegistry layoutProviderRegistry;
 
     private ArtifactByteStreamsCopyStrategy simpleArtifactByteStreams = SimpleArtifactByteStreamsCopy.INSTANCE;
-
+    
+    private ThreadLocal<ArtefactCopyContext> artefactCopyContext = new ThreadLocal<>();;
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java]: I have huge doubts changes in this class are related with `SB-930` :)

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java (hunk 3 above)
Comment (sbespalov) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java]: `org.carlspring.strongbox.configuration.Configuration#storages` is `@Embeded`.

File: strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java (hunk 2 above)
Comment (sbespalov) [reply strongbox-data-service/src/main/java/org/carlspring/strongbox/data/service/CommonCrudService.java]: yeah, it can be utility method in the future

File: strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserAccessModel.java (hunk 4 above)
Comment (sbespalov) [reply strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/security/UserAccessModel.java]: yeah, maybe
will try it without cascade

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/AccessModel.java (hunk 10 above)
Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/AccessModel.java]: `AccessModel` is `@Embedded` in `org.carlspring.strongbox.users.domain.User`

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java (hunk 11 above)
Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: OrientDB support this

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java]: yeah, it has a point to be

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java (hunk 14 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java]: yes, it's not related
I have fixed resource leaks here
also, I wanted to ask
why you implemented this with recursion, instead of just for loop?

**Comment by sbespalov** [comment]
@fuss86 @carlspring regarding entity relation cascases. I will check it within `SB-946`

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java (hunk 14 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java]: also, I meet this resource leak issue within this PR, so that's why it's here

**Comment by carlspring** [comment]
Thanks for these fixes! :)

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java (hunk 11 above)
Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: > Shouldn't AccessModel be Embeddable in this case ? 

oh, now I see what you mean, it's `javax.persistence.Embeddable` annotation. 
But it's seems that OrientDB works without this with just `javax.persistence.Embedded`.
So as I sayed it's supports this as it is.

Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: ok, seems that OrientDB should work even without `javax.persistence.Embedded`, and I have a feeling that it assume that type is embedded if the field is not annotated wtih `ManyToOne` or other relation type.

**Comment by sbespalov** [comment]
@fuss86 there will be upcoming PR with some changes in our enity model, so feel free to add your thougths there.  it's SB-946

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java (hunk 11 above)
Comment (fuss86) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: So we can get rid of `javax.persistence.Embedded` from our source code so there won't be any misunderstandings to use it or not ?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java (hunk 1 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java]: Could you please explain `why` we need such things ? Will we have to do similar things for other entities ?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java (hunk 14 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java]: Please let me know if you find some issues with my code and I'll fix it ASAP. I'm curious where was the leak out there.

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ProxyRepositoryArtifactByteStreamsCopy.java]: the issue was that you using recursion there and open new connection down in the calling stack, and not close previous. I have fixed this with `ThreadLocal` context, so connection and client now shared in current thread between method calls.

File: strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java (hunk 11 above)
Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: the thing is that `Embedded` types have no `rid`. they just embedded as a part of enclosing entity. in general it's just ORM sugar.

Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/java/org/carlspring/strongbox/users/domain/User.java]: I think it can be useful just as info that this types are just Embedded.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java]: I leave the comment there, this is needed to reset `path` value of `ArtifactEntry`.
